### ISSUE TITLE
no-unresolved peerDependencies option

### ISSUE THREAD

**Issue by rafgraph** [import-js/eslint-plugin-import#825]
Add `peerDependencies` option to `import/no-unresolved` rule.

Use case: developing a React module that has React as a `peerDependency`, and the path cannot be resolved because it's not in `node_modules` (development testing is done via `yarn link`).

Currently I'm using `'import/no-unresolved': ['error', { ignore: ['^react$'] }]`, but would prefer to have something like `'import/no-unresolved': ['error', { peerDependencies: true }]`

**Comment by ljharb** [comment]
Linking is only for dev; you still need `npm ls` to exit zero after just running `npm install`.

Typically with React (and every peer dependency) this is solved by adding it as both a peer dep and a dev dep, so that `npm install` installs it for dev, but it's a peer dep for consumers. Then, if you want to `npm link`/`yarn link` on top of that, you can do so.

**Comment by rafrex** [comment]
That is true about `npm install`, it gives a warning that the peer dependency wasn't found (`npm ls` gives an error). However with `yarn install` and `yarn list` no such warning or error is given.

Adding `react` as a dev dependency while developing modules with linking is adding it solely for moral support as the `react` package in `node_modules` is never needed. I personally prefer to take the warning (or just use yarn) than to add dev dependencies for moral support reasons.

I'm suggesting making it a non-default option that can be enabled to avoid adding unused dev dependencies.

**Comment by ljharb** [comment]
It's not an unused dependency; if it's not in package.json, it doesn't belong in your node_modules directory - linking should only shadow things that `npm install` would put there. `yarn install` shouldn't differ from `npm install` except in how versions and subdeps are resolved.

**Comment by rafrex** [comment]
It is an unused dependency if you set `resolve: { symlinks: false }` in the app's webpack config so you don't end up using two different versions of the peer dependency (e.g. React).

Personally I think it's bad practice to follow symlinks for peer dependencies for this very reason. In development (with linking) your module is using the React version (or some other peer dependency) from it's `node_modules` folder while your app is using the React version from it's `node_modules` folder, but when you stop linking they're both going to use the React version in the app's `node_modules` folder, so IMO it's better to develop that way too.

For example see [React Interactive](https://github.com/rafrex/react-interactive) - the app is in the `gh-pages` branch and the component is in the `master` branch.

**Comment by rafrex** [comment]
In the app's webpack config you can also tell it to explicitly use it's own `node_modules` instead of setting `symlinks: false` by adding `resolve: { modules: [path.resolve(__dirname, 'node_modules')] }`. They both have the same result when developing with linking.

**Comment by rafrex** [comment]
It depends on the app's webpack config settings:

Not used as long as the app has the peer dependency installed:
`resolve: { modules: [path.resolve(__dirname, 'node_modules'), 'node_modules'] }`

Used only if it is present in the module's dev dependencies (along with a peer dependency), but is not required to be a dev dependency (if it's not found in the module's `node_modules` directory it looks in the app's `node_modules` directory):
`resolve: { symlinks: false }`
OR
`resolve: { modules: ['node_modules', path.resolve(__dirname, 'node_modules')] }`

I prefer to use `resolve: { symlinks: false }` without listing it as a dev dependency because then for regular dependencies it looks in the module's `node_modules` directory first before the app's and acts much like it is installed normally (without linking), the only difference is it doesn't de-dupe for shared regular dependencies.

Personally I think it's bad practice to add dev dependencies for peer dependencies for this very reason. In development (with linking) your module is using the React version (or some other peer dependency) from it's `node_modules` directory (put there by the dev dependency) while your app is using the React version from it's `node_modules` directory, but when you stop linking they're both going to use the React version in the app's `node_modules` folder, so IMO it's better to develop that way too.

For example see [React Interactive](https://github.com/rafrex/react-interactive) - the app is in the `gh-pages` branch and the component is in the `master` branch.

**Comment by ljharb** [comment]
I realize that linking can create these complications, but one has to manually run a number of commands to keep linked repos up to date, which could include installing specific versions of react. In practice this doesn't turn out to be a problem, in my experience.

**Comment by rafrex** [comment]
Fair enough, but why not add a non-default option for `peerDependencies` to `import/no-unresolved` for those who have a different approach?

(I've added a few basic npm scripts to my workflow that make the linking and unlinking process fairly automatic).

**Comment by ljharb** [comment]
If the link is set up properly, then wouldn't the resolver be able to locate React, such that it's not unresolved? I'm confused how `node_modules/react` doesn't exist, but your code still works.

You may need a custom resolver (that matches your webpack logic), sure, but that shouldn't require a change to the rule.

**Comment by rafrex** [comment]
```
App 
|--node_modules
|  |--react (as a dep of app)
|  |--my-component (as a symlink, which has react as a peer dep)
|  |  |--node_modules (doesn't contain react)
|  |  |  |--some-dep-of-my-component
|--webpack.config.js (with resolve: { symlinks: false })
```
With webpack not following symlinks it simply looks up the tree as though the symlink is a regular directory and finds `react` in the App's `node_modules` so using `my-component` in the app works fine. 

The actual location of the `my-component` directory relative to the `app` directory is arbitrary and in theory `my-component` could be linked to multiple apps at the same time.

The issue is in `my-component` where it has `react` as a peer dependency and there is no way to know what `app(s)` it's been linked to, if any, to resolve the peer dependency (at least that I'm aware of). Note that `my-component` is useless outside of an app and won't be run on it's own (hence react is a peer dependency).

To see how this works in practice, check out [React Interactive](https://github.com/rafrex/react-interactive) - the app is in the `gh-pages` branch and the component is in the `master` branch. Clone the branches in two different directories and run `npm run dev` in the component directory and `npm run start-local` in the app. I also created a boilerplate based this concept, see [react-component-dev](https://github.com/rafrex/react-component-dev).

**Comment by rafrex** [comment]
@ljharb just following up on this, any thoughts?

**Comment by ljharb** [comment]
When you run `require` in the app in your above diagram, `my-component`'s node_modules that `require('react')` would walk upwards, and use the top-level react, so I'm not seeing the problem.

**Comment by rafrex** [comment]
Nope. Eslint follows the symlink so it doesn't see the directory structure like the above diagram. Only the webpack resolver doesn't follow the symlink because of `resolve: { symlinks: false }` in the webpack config.

Is there a way to have eslint not follow symlinks? If there is, then it would see the directory structure the same as webpack and that would work.

**Comment by rafrex** [comment]
The issue is not in `app`. It is in `my-component` which is in a totally different directory and has no awareness of being in the `app`'s `node_modules`.

**Comment by rafrex** [comment]
`my-component` has the `peerDependency` of `react`

**Comment by ljharb** [comment]
Ah - you're saying the issue arises when App and `my-component` each need their own copy of React in order to run the build processes in each.

The solution here is that linking `my-component` necessitates also linking *every peer dep* of `my-component`. Thus, you need to `npm link` one of the React copies into the other one - that way it will share the same version of React.

This is just the nature of peer deps combined with the nature of `npm link`. A linting rule change won't solve your problems; but linking `react` in both places will.

**Comment by rafrex** [comment]
Not exactly... the code in `my-component` is never run outside of *an* app, and `my-component` can be linked to multiple apps (think of it as one to many). So it doesn't make sense to link from the app back to `my-component`.

Everything runs fine the way it is. The only issue is the `import/no-unresolved` rule in `my-component` gives me an error because the `peerDependency` `react` isn't in `my-component`'s `node_modules`.

So having something like `'import/no-unresolved': ['error', { peerDependencies: true }]` is pretty much the only way to solve the issue.

**Comment by ljharb** [comment]
it's not "from the app" - `npm link` uses `npm root -g` as a global passthrough. One alternative is to clone the react github repo and `npm link` from there, and then `npm link react` into every project on your machine that uses React.

In general, while `my-component` is linked, you shouldn't be running the linter there - the linter is what you run after `rm -rf node_modules && npm install`.

**Comment by rafrex** [comment]
To me it doesn't make sense to link react that way - seems like uneccasry complexity.

my-component doesn't need its own copy of react for its build process (which consists of only running babel).

I use a linter setup in atom so it's runs as I type.

The error is providing no useful feedback with this type of dev setup and I don't see why an option can't be added to the rule? Different people have different opinions of best practices, and can set the rule for what they see as best practices.

**Comment by rafrex** [comment]
@ljharb I'm confused why this option can't be added?

In `my-component` with `react` as a `peerDependency`, `react` is not needed and is not present in `my-component`'s node_modules and everything works fine using the dev setup I explained above.

**Comment by ljharb** [comment]
Again, while my-component is linked, caveats apply to developing in it - including, you just need to ignore the linting errors.

The only use case you've demonstrated is `npm link` - the far more common use case is linting a project as part of CI, to gate a PR or a publish. Creating this option will enable you to configure your linter so that it's silent during `link` development - a time when you can easily just choose to ignore the warnings - and as a result, you might end up publishing with an error in your dependencies.

**Comment by rafrex** [comment]
But isn't that for me to decide how to configure the linter for my needs?
Is what you're saying is that you don't want to support linking development because other people might mis-configure the linter for there needs?

**Comment by rafrex** [comment]
cc @benmosher

**Comment by ljharb** [comment]
I'm saying that the use case for a rule - something that gets committed to a repo - needs to be stronger than just "only for npm link".

In the link case, you can always safely choose to ignore the warning - indeed, to never run the linter at all. That's not a safe choice in the non-link case.

**Comment by rafrex** [comment]
🤔😞 I guess I'll just keep using the workaround
`'import/no-unresolved': ['error', { ignore: ['^react$'] }]`

**Comment by macrozone** [comment]
I have the same issue as @rafrex 

also installing react in "my-component" might result in a "two versions of react included" problem on the project where I `yarn link`ed the local "my-component". 

So 'import/no-unresolved': ['error', { peerDependencies: true }] seems to be usefull to me

**Comment by benmosher** [comment]
@rafrex doesn't `yarn link` put a `react` symlink to wherever in your `node_modules`?

sounds like a bug where `no-unresolved` doesn't respect symlinks? I would think it shouldn't matter but from what you've reported, sounds like it does.

**Comment by macrozone** [comment]
@benmosher `yarn link my-component` in the app project will add a symlink inside node_modules to your local my-component directory. If the node_modules of "my-component" contains a `react` version as well, "my-component" seems to pull that one in, resulting in a possible "two versions of react included" error.

**Comment by benmosher** [comment]
Sure, so what I don't understand is: given the following

> Use case: developing a React module that has React as a `peerDependency`, and the path cannot be resolved because it's not in `node_modules` (development testing is done via `yarn link`).

the problems seems it must be one of either:

- `no-unresolved` doesn't follow symlinks (would be both a surprise and a major bug)
- OP hasn't yet `yarn link`'d at lint time

for which the resolutions seem to be:

- fix symlink bug, or
- `yarn link` before lint time.


oooooh you know what, I think I finally get it. y'all aren't `yarn link`-ing `react`, you're `yarn link`-ing your project with the peer dependency into some test environment where `react` is also installed?

**Comment by ljharb** [comment]
Yes, that's the situation - they're linting inside the linked-out project while developing it.

**Comment by rafrex** [comment]
Yes, that's the use case.

**Comment by benmosher** [comment]
I would be inclined to make that an option implemented in the Node resolver itself instead of in the rule, since it is an artifact of the Node/npm world and not a JS feature. 

If enabled, any such setting would instruct the Node resolver to check package.json and report that modules exist if they match a peer dependency, not unlike the existing `code-modules` setting.

**Comment by ludovicmnji** [comment]
I'm in the same use case of @rafrex and I have the same issue. I don't know how to deal with peerDependencies (which cannot be included in devDependencies like `react` or `styled-components`). 

The only solution I have for the moment is to list all my peerDependencies and ignore them in `.eslintrc`. Is it the solution you recommend finally or there was some new since last year ?

**Comment by ljharb** [comment]
@ludovicmnji in fact, everything that is a peer dep *must* be also included in either deps or dev deps, including things like react.

**Comment by ludovicmnji** [comment]
@ljharb I'm agree with you, and it was the naive approach that I have when I started : I added all my peerDependencies in devDependencies (it seems logical). But when I did this, I had very weird issues with styled-components and React when I was building the project who use the library with peerDependencies & devDependencies (duplicate stories).

Maybe I should ask directly to React team or styled-components contributors if it's a normal behaviour belong to them ... 

Thanks for your reply (and your confirmation).

**Comment by ljharb** [comment]
I have tons of projects using React this way and it works fine; I’ve never used styled components tho.

**Comment by menocomp** [comment]
I has the same issue. 
As a workaround I am using https://www.npmjs.com/package/install-peers to install peer dependencies and let the linter succeed.

**Comment by lorenzomigliorero** [comment]
This seems to be a good compromise for me:

```
const { peerDependencies } = require('./package.json');
module.exports = {
    "rules": {
      "import/no-unresolved": ["error", { ignore: Object.keys(peerDependencies) }]
    }
};
```

**Comment by ljharb** [comment]
@lorenzomigliorero can you elaborate on why any of your peer deps aren't also in either your deps or dev deps, so that they're always locally installed?

**Comment by ivdma** [comment]
Maybe I can add my ¢2,- here.

So I have a UI Library called `ui-components`. This library is using React amongst dozens of other packages. React wants us to have only one copy of React library installed in the project, which became notable after they introduced hooks (React 16.8).

So my main project which includes `ui-components`, complains whenever there are multiple React packages are present in the build. The second React package reference in package-lock.json comes from `ui-components`. I've tried to remove react and react-dom from both `dependencies` and `devDependencies` blocks in package.json and it fixed the error. The error itself, by the way, looks like this and includes a stack which does not say a lot about the actual error (I removed the stack, but if you like, I can add it):

```
Error: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
```

I could easily check numbers 1 and 2: 1) versions of react and a react renderer are equal; and 2) we didn't use any hooks in the main app (except for packages we included in the main app). So maybe it's number 3. Like I mentioned, I removed react and react-dom from `dependencies` and `devDependencies`. After linking, the above error in the main project was gone, but it introduced linter errors in `ui-components` which I, like @rafrex and others in this issue, tried to turn off.

So I started digging and came across this issue. 

> @lorenzomigliorero can you elaborate on why any of your peer deps aren't also in either your deps or dev deps, so that they're always locally installed?

So basically, when I add `react` and `react-dom` to either `dependencies` or `devDependencies` blocks, another react package gets included in the main app's bundle despite it being present in `peerDependencies` block. 

So that's why the need for `{ peerDependencies: true }` option. 

However I realize that this solution might fix the symptoms of a totally different problem, but at this moment the only thing I know for sure is that removing react and react-dom from `dependencies` and `devDependencies` fixes my problem.

**Comment by ljharb** [comment]
@ivdma the solution is for ui-components to have react as both a peer dep and a dev dep.

**Comment by ivdma** [comment]
Yes and no. 
Imagine this scenario:

```bash
cd ~/projects/ui-components
npm ci
npm link

cd ~/projects/main-project
npm ci
npm link ui-components
```

At that moment there are (at least) two copies of react in `main-project/node_modules` which confuses webpack.

This comment (hidden among hundreds of other comments) was actually useful and this is how I fixed in the main project: https://github.com/facebook/react/issues/13991#issuecomment-463486871

So no: even if I marked `react` and `react-dom` as peerDependencies in `ui-components` they still were present in the main project. Maybe worth mentioning: there's no webpack or other bundler in `ui-components`. We only use babel to transpile ES6 and JSX from `src/` to plain old JS in `dist/`.

**Comment by ljharb** [comment]
When using npm link, everything that’s a peer dep must *also*, separately, be linked.

**Comment by ivdma** [comment]
> When using npm link, everything that’s a peer dep must _also_, separately, be linked.

How does that work?

Tried this:
```
cd ~/projects/ui-components # put react and react-dom to peer- and devDependencies here
npm ci
npm link

cd ~/projects/main-project
npm ci
npm link ui-components
npm link react
npm link react-dom
```

But still no luck.

**Comment by ljharb** [comment]
You have to have react and react-dom in their own repos, also linked.

**Comment by Anenth** [comment]
I fixed this issue by adding `no-extraneous-dependencies` rule and removed the `no-unresolved` rule
https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-extraneous-dependencies.md

```diff
- 'import/no-unresolved': 2,
+ 'import/no-extraneous-dependencies': ['error', { devDependencies: false, peerDependencies: true }],
```

**Comment by cha0s** [comment]
tl;dr A library/UMD bundle with external imports should not require physical presence of the external module at build/lint time, especially when architectural semantics ensure that the external dependency will be available at application build.

---

I have the same issue. I develop plugins for a plugin system. The plugins implement hooks named e.g. `thing/hook-name`.

Due to the semantics of the application and plugin system, **`thing` hooks will never be invoked by the application if `thing` is not already `require`able (...and, `require`d) in the main application context build/runtime. Also, **all imports are considered external at library build-time**. I don't bundle React with my plugin that implements a React hook. It's marked as a webpack external and then at application build, React is made available to any plugin that has a reference to an external React.

I create a new plugin called `foobar`. `foobar` implements a `thing` hook. So this means that at runtime, `thing` is guaranteed to be available (due to the aforementioned semantics of my application/plugin architecture). 

The `foobar` developer should list `thing` in `peerDependencies`. That way we correctly identify that `thing` is not a true dependency (or even dev dependency) of `foobar`, but that it will exist as a 'peer' at application build/runtime.

I hope this use case isn't too confusing to understand, but I assure you that there is reason to want this feature beyond some specific yarn/linking semantics. It's of course possible to specify these as 'hard' dependencies, but it's just polluting `node_modules` and it smells.

---

I just wanted to try to convince you that there are legitimate usecases for this feature. https://github.com/benmosher/eslint-plugin-import/issues/825#issuecomment-542618188 is what I will be going with to make this automatically work in practice, but I think you have to agree that it's a hack.

**Comment by ljharb** [comment]
If your peer dep is *also* a devDep, it would be installed without breaking anything for your consumers.

How can you possibly test your plugin (inside the repo, which is the only place it's really important) if you don't have the plugin system installed?

**Comment by cha0s** [comment]
Nothing is broken for my consumers. The only thing broken is the linter, reporting problems that don't exist. :)

If I mark it as a dev dependency then it gets installed into my plugin's `node_modules` directory. This is wasteful and unnecessary.

**The plugin build process does not need the peer dependency at build-time because it is marked as external**. I am not bundling the peer dependency with my code.

Testing the plugin vis-a-vis the plugin architecture doesn't make a lot of sense. I have tests for the plugin architecture itself, and tests of whatever functionality a plugin may export. In the rare case of e.g. a plugin modifying/decorating/whatevering other plugins' outputs, then I can mock limited parts of the system.

Also I do feel like you are fundamentally misunderstanding the point here. It's not about having the plugin system installed.

I want you to imagine (it's real but it isn't open source code) that I have a magical architecture where I can dynamically build an application by simply specifying a list of package names which compose the application. The architecture invokes a hook called `appStart`, and any of the packages that I specified in my magical list of packages can listen for that hook and then do whatever.

Now say I have two packages, `foo` and `bar`.

`foo` exports a function, `baz`. What `baz` does doesn't matter. Foo also exports a listener for the `appStart` hook. Now for the 'do whatever' part: when `foo`'s `appStart` hook implementation is called, `foo` then uses the plugin architecture to invoke another hook: `fooStart`. I didn't tell you before, but that's exactly what `bar` is listening for: `fooStart`. When `bar` gets this `fooStart` hook, `bar` then calls `foo.baz();`. In other words, `bar` has a peer dependency on foo. **There is no way that `fooStart` can be invoked if foo isn't already `require`d by the main (magical) application.** `bar` *does not need` foo at build-time. `bar` simply marks `foo` as external, and has access to the actual `foo` when the application runs.

Does this make sense?

**Comment by ljharb** [comment]
"Wasteful" isn't really important; disk space is infinite and free.

I do think it makes sense to test the plugin system inside the plugin itself - that's *precisely* what this plugin does; it tests itself with eslint. If the only way you test the plugin is outside the repo, you've created a lot of friction to prove that your changes work. In your example, your tests *inside the plugin repo* should include that magical application.

The point is that peer deps MUST ALWAYS be installed, without exception, for every use case, full stop.

**Comment by cha0s** [comment]
> "Wasteful" isn't really important; disk space is infinite and free.

Seems pretty subjective.

> I do think it makes sense to test the plugin system inside the plugin itself - that's precisely what this plugin does; it tests itself with eslint. If the only way you test the plugin is outside the repo, you've created a lot of friction to prove that your changes work. In your example, your tests inside the plugin repo should include that magical application.

You have completely missed the point, again. **The "magical application" isn't the peer dependency**. I don't know why you are so fixated on testing semantics to begin with.

> The point is that peer deps MUST ALWAYS be installed, without exception, for every use case, full stop.

The takeaway from this ticket is that you are not going to reponsible for this fix. Got it.

**Comment by soryy708** [comment]
@ljharb is there anything actionable for `eslint-plugin-import` here? it looks like the conversation shifted from a rule enhancement to project config and package management etc, so it's unclear what would a contributor do with this issue.

**Comment by ljharb** [comment]
Yeah, I think this is just a many-year misunderstanding of how peer deps work, how linking works, and how the two work (or don't work) in concert.
